### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by aait**
This PR contain such improvements like:
* Subscriptions fixes for cards and mobile pay
* Payment status checking on "Thank you" page 
* Business logic has been moved to separate files
* and other multiple changes...

### CODE REVIEW

**Comment by antondrob** [comment]
Test

File: includes/class-wc-reepay-admin.php [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+			wp_enqueue_script( 'reepay-admin-js' );
+		}
+	}
+
+	/**
+	 * Action for Capture
+	 */
+	public function ajax_reepay_capture() {
+		if ( ! wp_verify_nonce( $_REQUEST['nonce'], 'reepay' ) ) {
+			exit( 'No naughty business' );
```
Comment (antondrob) [inline includes/class-wc-reepay-admin.php:243]: Test

Comment (dimaspolohov) [inline includes/class-wc-reepay-admin.php:243]: When closing the popup and canceling the order, users are redirected to an empty page with a strange inscription “No naughty business” https://prnt.sc/VFF-3A2kMi3A, this text is also displayed in the code in 5 more places https://prnt.sc/KGHSR8HkxROZ, they are displayed when the nonce check fails, but during the test I was transferred to such a page

File: includes/abstracts/abstract-wc-gateway-reepay.php [hunk 2]
```diff
@@ -168,38 +155,154 @@ public function __construct() {
...
+	 * @return void
+	 */
+	public function refund_payment( $order, $amount = false, $reason = '' ) {
+		if ( is_int( $order ) ) {
+			$order = wc_get_order( $order );
+		}
+
+		// Check if the order is cancelled - if so - then return as nothing has happened
+		if ( '1' === $order->get_meta( '_reepay_order_cancelled', true ) ) {
+			return;
```
[the remark this answers is not in the document]
Comment (markusbrunke) [reply includes/abstracts/abstract-wc-gateway-reepay.php:260]: @aait Will you take a look?

File: includes/class-wc-reepay-thankyou.php [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+		if ( ! in_array( $order->get_payment_method(), WC_ReepayCheckout::PAYMENT_METHODS, true ) ) {
+			return;
+		}
+
+		$gateway = rp_get_payment_method( $order );
+
+		// Complete payment if zero amount
+		if ( abs( $order->get_total() ) < 0.01 ) {
+			$order->payment_complete();
```
[the remark this answers is not in the document]
Comment (markusbrunke) [reply includes/class-wc-reepay-thankyou.php:82]: @aait Will you take a look?

[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-thankyou.php:89]: This line is required for Subscriptions. When customer is changing a payment method to another ('Payment method switch').
Or when order total is zero when coupon code / discount was applied for 100%.

File: includes/abstracts/abstract-wc-gateway-reepay.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (aait) [reply includes/abstracts/abstract-wc-gateway-reepay.php]: Fixed

File: includes/class-wc-reepay-admin.php [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+			// Enqueued script with localized data
+			wp_enqueue_script( 'reepay-admin-js' );
+		}
+	}
+
+	/**
+	 * Action for Capture
+	 */
+	public function ajax_reepay_capture() {
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-admin.php:241]: Fixed

File: assets/js/admin.js [hunk 5]
```diff
@@ -181,14 +181,13 @@ jQuery(document).ready(function ($) {
 		var formatted_amount = $("#reepay_order_total").val();
```
[the remark this answers is not in the document]
Comment (aait) [reply assets/js/admin.js:179]: Fixed

File: includes/class-wc-reepay-api.php [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+			$order = wc_get_order( $order );
+		}
+
+		if ( ! in_array( $order->get_payment_method(), WC_ReepayCheckout::PAYMENT_METHODS, true ) ) {
+			return new WP_Error( 0, 'Unable to get invoice data.' );
+		}
+
+		$order_data = $this->get_invoice_by_handle( rp_get_order_handle( $order ) );
+		if ( is_wp_error( $order_data ) ) {
+			/** @var WP_Error $order_data */
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-api.php:159]: Fixed

File: includes/class-wc-reepay-api.php [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+			$order = wc_get_order( $order );
+		}
+
+		if ( ! $amount ) {
+			$amount = $order->get_total();
+		}
+
+		$result = $this->get_invoice_data( $order );
+		if ( is_wp_error( $result ) ) {
+			/** @var WP_Error $result */
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-api.php:198]: Fixed

File: includes/class-wc-reepay-api.php [hunk 8]
```diff
@@ -0,0 +1,772 @@
...
+		if ( is_int( $order ) ) {
+			$order = wc_get_order( $order );
+		}
+
+		if ( ! $amount ) {
+			$amount = $order->get_total();
+		}
+
+		if ( ! $this->can_capture( $order, $amount ) ) {
+			return new WP_Error( 0, 'Payment can\'t be captured.' );
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-api.php:311]: Fixed

File: includes/class-wc-reepay-api.php [hunk 9]
```diff
@@ -0,0 +1,772 @@
...
+			'handle'    => rp_get_order_handle( $order ),
+			'amount'    => rp_prepare_amount( $amount, $currency ),
+			'currency'  => $currency,
+			'source'    => $token,
+			'recurring' => order_contains_subscription( $order ),
+		);
+		$result = $this->request( 'POST', 'https://api.reepay.com/v1/charge', $params );
+		if ( ! is_wp_error( $result ) ) {
+			$this->process_charge_result( $order, $result );
+
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-api.php:368]: Fixed

File: includes/class-wc-reepay-api.php [hunk 10]
```diff
@@ -0,0 +1,772 @@
...
+		     ( in_array( $result->get_error_code(), [ 105, 79, 29, 99, 72 ] ) )
+		) {
+			// Workaround: handle already exists lets create another with unique handle
+			$params['handle'] = rp_get_order_handle( $order, true );
+			$result           = $this->request( 'POST', 'https://api.reepay.com/v1/charge', $params );
+			if ( ! is_wp_error( $result ) ) {
+				$this->process_charge_result( $order, $result );
+
+				return $result;
+			}
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/class-wc-reepay-api.php:388]: Fixed

File: includes/functions.php [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+	 * Get Order By Reepay Order Handle.
+	 *
+	 * @param string $handle
+	 *
+	 * @return false|WC_Order
+	 */
+	function rp_get_order_by_handle( $handle ) {
+		global $wpdb;
+
+		$query    = "
```
[the remark this answers is not in the document]
Comment (aait) [reply includes/functions.php]: Fixed using `wc_get_orders()`

File: includes/class-wc-reepay-admin.php (hunk 1 above)
Comment (aait) [reply includes/class-wc-reepay-admin.php:243]: Fixed

File: includes/class-wc-reepay-api.php [hunk 12]
```diff
@@ -318,37 +365,46 @@ public function charge( WC_Order $order, $token, $amount, $currency ) {
...
+			$result = $this->request( 'POST', 'https://api.reepay.com/v1/charge', $params );
+			if ( is_wp_error( $result ) ) {
+				if ( 'yes' == $this->gateway->handle_failover &&
+				     ( in_array( $result->get_error_code(), array( 105, 79, 29, 99, 72 ) ) )
+				) {
+					// Workaround: handle already exists lets create another with unique handle
+					$params['handle'] = rp_get_order_handle( $order, true );
+					$result           = $this->request( 'POST', 'https://api.reepay.com/v1/charge', $params );
+					if ( is_wp_error( $result ) ) {
+						throw new Exception( $result->get_error_message(), $result->get_error_code() );
```
Comment (dimaspolohov) [inline includes/class-wc-reepay-api.php:379]: I think here need to add logging

File: includes/class-wc-reepay-api.php [hunk 13]
```diff
@@ -318,37 +365,46 @@ public function charge( WC_Order $order, $token, $amount, $currency ) {
...
-			// Workaround: handle already exists lets create another with unique handle
-			$params['handle'] = rp_get_order_handle( $order, true );
-			$result           = $this->request( 'POST', 'https://api.reepay.com/v1/charge', $params );
-			if ( ! is_wp_error( $result ) ) {
-				$this->process_charge_result( $order, $result );
+					return $result;
+				}
 
-				return $result;
+				throw new Exception( $result->get_error_message(), $result->get_error_code() );
```
Comment (dimaspolohov) [inline includes/class-wc-reepay-api.php:387]: I think here need to add logging

**Pull request merged**